### PR TITLE
Suggest cloning `Arc` moved into closure

### LINKED ISSUE #104232: Suggest cloning Arc/Rc
```rust
fn main() {
    let x = Arc::new(1);
    foo(x);
    bar(x);
}

fn foo(_: Arc<usize>) {}
fn bar(_: Arc<usize>) {}
```
```
error[[E0382]](https://doc.rust-lang.org/stable/error-index.html#E0382): use of moved value: `x`
 --> src/main.rs:6:9
  |
4 |     let x = Arc::new(1);
  |         - move occurs because `x` has type `std::sync::Arc<usize>`, which does not implement the `Copy` trait
5 |     foo(x);
  |         - value moved here
6 |     bar(x);
  |         ^ value used here after move
  ```

The help message could mention that `Arc` implements `Clone`. We don't usually suggest cloning values in diagnostics, but for `Arc/Rc` we probably should.

Another common case (this one is probably harder to suggest a fix for):
```rust
fn main() {
    let x = Arc::new(1);
    for _ in 0..4 {
        std::thread::spawn(move || {
            println!("{}", x);
        });
    }
}
```
```
error[[E0382]](https://doc.rust-lang.org/stable/error-index.html#E0382): use of moved value: `x`
 --> src/main.rs:6:28
  |
4 |     let x = Arc::new(1);
  |         - move occurs because `x` has type `std::sync::Arc<i32>`, which does not implement the `Copy` trait
5 |     for _ in 0..4 {
6 |         std::thread::spawn(move || {
  |                            ^^^^^^^ value moved into closure here, in previous iteration of loop
7 |             println!("{}", x);
  |                            - use occurs due to use in closure
```

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0382]: borrow of moved value: `x`
  --> $DIR/moves-based-on-type-capture-clause-bad.rs:9:20
   |
LL |     let x = "Hello world!".to_string();
   |         - move occurs because `x` has type `String`, which does not implement the `Copy` trait
LL |     thread::spawn(move || {
   |                   ------- value moved into closure here
LL |         println!("{}", x);
   |                        - variable moved due to use in closure
LL |     });
LL |     println!("{}", x);
   |                    ^ value borrowed here after move
   |
   = note: this error originates in the macro `$crate::format_args_nl` which comes from the expansion of the macro `println` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider cloning the value before moving it into the closure
   |
LL ~     let value = x.clone();
LL ~     thread::spawn(move || {
LL ~         println!("{}", value);
   |
```

Fix rust-lang/rust#104232.

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? compiler

too busy to review this

**Comment by pnkfelix** [comment]
@estebank , isn't this suggesting cloning a *lot* more stuff than just `Arc`/`Rc` ?

The description for issue #104232 explicitly said:

> ﻿We don't usually suggest cloning values in diagnostics, but for `Arc`/`Rc` we probably should.

which I took to mean that we *wouldn't* jump to suggesting cloning for things like `Box` nor `String`. But it looks to me like in this PR, you are indeed suggesting clones for `Box` and `String` ?

**Comment by pnkfelix** [comment]
@rfcbot author

**Comment by estebank** [comment]
> The underling assumption of https://github.com/rust-lang/rust/issues/104232 is that cloning an Rc is quite cheap, and cloning an Arc is potentially also cheap (apart from the potential for having many threads contending to increment the reference-count). But the cost of cloning a Box or String is dependent on their associated payload.)

The thing is that given the case being checked for here, either people `clone`, or they completely rework the code they have. I know this PR covers strictly more than was requested in the ticket, but I've been leaning lately towards the premise that telling people they can clone when they can is rarely a bad idea, despite the potential costs. I can modify this to rely on an allow-list, but even then I would still include `Box` and `String` as "can suggest".

**Comment by alex-semenyuk** [comment]
Ping from triage.
@pnkfelix Are we interesting in this PR since @estebank answered on your comment? if so @estebank should update it to fix conflicts

**Comment by pnkfelix** [comment]
I don't have further review commentary.

I remain personally skeptical about increasing the scope of the suggestion here, but we can land this and see what kinds of behaviors result. I.e. I am willing to be more liberal with diagnostics like this, since we can go back and reduce their scope later after we get more data about their impact.

@rustbot author

**Comment by Dylan-DPC** [comment]
r? compiler

**Review by davidtwco — APPROVED** [review]
I don't mind this suggestion, I might suggest rewording it to be more of a suggestion, "you could consider cloning", but I think it's an improvement, r=me after rebasing w/ or w/out wording change

**Comment by compiler-errors** [comment]
diagnostics only

@bors rollup
